### PR TITLE
Flush after scheduled time period

### PR DESCRIPTION
**Pull request by afoerster**
Fixes issue where messages could have been left in the buffer for a long
    period of time if not flushed out by another log message.

### CODE REVIEW

File: log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java [hunk 1]
```diff
@@ -96,28 +109,44 @@ private void flush() throws Exception {
     }
   }
 
+  /**
+   * Flush only if conditions are met
+   *
+   * @throws Exception
+   */
+  void tryFlush() throws Exception {
```
Comment (brockn) [inline log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: I don't see this being used

File: log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java [hunk 2]
```diff
@@ -36,6 +39,8 @@
 public class HttpAppender extends AppenderSkeleton {
 
   public static final String HOSTNAME = "hostname";
+  public final long FLUSH_PERIOD_SECONDS = 3;
```
Comment (brockn) [inline log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: might as well make it static

File: log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java [hunk 3]
```diff
@@ -49,6 +54,7 @@
   private long backoffTimeSeconds = INITIAL_BACKOFF_TIME_SECONDS;
 
   private String hostname = null;
+  ScheduledFuture scheduledFlushTask = null;
```
Comment (brockn) [inline log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: private final

File: log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java [hunk 4]
```diff
@@ -62,30 +68,37 @@ public void run() {
...
     }
   }
 
   /**
    * Flush all log events
+   *
    * @throws Exception
    */
-  private void flush() throws Exception {
+  void flush() throws Exception {
```
Comment (brockn) [inline log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: This method is now called in parallel. Some of these variables are not protected. Is that ok?

Comment (bpmcd) [reply log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: Maybe let the FlusherTask take care of calling this method, and remove the call from '.append()' to immediately flush?  Dunno, just an idea.

File: log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java (hunk 1 above)
Comment (bpmcd) [reply log-appender/src/main/java/io/phdata/pulse/log/HttpAppender.java]: If this method gets removed, you can also delete 'shouldFlush()' as it is only used here.

**Comment by afoerster** [comment]
I tried to do an ammend and force push on the branch but it looks like that blew away your comments. Also not sure why it's saying there are two commits to be merged including 'Always flush buffers on an ERROR message ', this was merged in before. If the commit looks good I'll take care of that in the merge.

The current behavior for flushing
- flush on scheduled interval if we aren't in an exponential backoff
- flush no matter what on close
- flush on ERROR message if we aren't in exponential backoff
- flush when message buffer exceeds max size

**Comment by afoerster** [comment]
> Maybe let the FlusherTask take care of calling this method, and remove the call from '.append()' to immediately flush? Dunno, just an idea.
I still left the check for `flush()` inside the append, can maybe remove it though.

**Comment by brockn** [comment]
Lgtm

**Pull request merged**